Adds capability to provision directories on the EFS dynamically

jonathanrainer opened the pull request:
This PR was the result of a hackathon at my current company as well as some work outside to sure up the unit tests and E2E test the final product.

**Is this a bug fix or adding new feature?**
New feature, requested in #538 and #517 and possibly other issues, this comes up a lot as a feature people would like.

**What is this PR about? / Why do we need it?**
Since its creation the driver has supported Access Point Provisioning as its main means of dynamic provisioning. However this is problematic for a few reasons, it can causes issues with user permissions around deleting provisioned directories and there is a hard limit of 120 AccessPoints per EFS which for some use cases is very quickly depleted. Further it requires various AWS IAM Permissions that can be complicated to sort out and manage.

This PR allows a new provisioning mode called `efs-dir` so that instead of creating EFS access points, directories are created instead. This new provisioning mode supports all of the previous parameters (UID, GID, GID Ranges etc.) and so is very easy to transition to, or run alongside the access point provisioning style if preferred. At present the directories are named after the PersistentVolume that is created to ensure a unique directory name and not run into security problems but once #640 merges there's no reason why the same method wouldn't apply to this provisioning method as well with a little bit of extra work. Either this PR can be updated or #640 can be extended if this merges first.

This is achieved by creating a new Interface called Provisioner that is implemented by an AccessPointProvisioner (the original method) and a DirectoryProvisioner (the new method) this also allows in future for different kinds of provisioning to occur, maybe FileSystemProvisioning for example. 

**What testing is done?** 
I've moved around some of the Unit Tests that used to relate to `controller.go` into new files that reference each of the provisioners. These all run 🟢 and I'm fairly happy I've covered most of the important cases. Overall I'm not 100% sure about the placement of the tests and the coverage but it definitely covers all the new code and the code that existed and replicates existing behaviour.

I've also run this through on my own EKS cluster with the `deleteProvisionedDir` flag on and off and it performs exactly as I anticipated it would. Directories get provisioned, and then when the PVC gets deleted they either are deleted or remain. I've also written E2E tests to run in CI for creation and deletion but will need the PR to be approved before I can road test those.

fixes #538

jonathanrainer commented:
@Ashley-wenyizha Hi, could I get an approved to test on here so I can start sorting out the E2E tests?

thesuperzapper commented:
@Ashley-wenyizha @wongma7 @jsafrane I truly believe a feature like this PR is critical for the EFS CSI driver to remain useful, and I would greatly appreciate your thoughts on this PR.

With the current approach, people will often get themselves into a situation where they are suddenly unable to provision new PVCs (once they hit 120 EFS "Access Points" in an AWS Account) and then have no idea what is wrong, let alone be able to fix it without moving to another CSI driver.

The [`kubernetes-csi/csi-driver-nfs`](https://github.com/kubernetes-csi/csi-driver-nfs) driver already uses a "sub-folder" approach that works on arbitrary NFS servers (including EFS), so if the EFS driver is unwilling to implement this feature, we should consider __deprecating this EFS driver__ and point people towards the official NFS one, in the interest of not leaving users with a ticking timebomb that breaks their cluster once 120 PVCs exist.

jonathanrainer commented on pkg/driver/directory_provisioner.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+func (d DirectoryProvisioner) Delete(ctx context.Context, req *csi.DeleteVolumeRequest) error {
+	if d.deleteProvisionedDir {
```
Done, I can see it's a much better way to express what I was going for

jonathanrainer commented on pkg/driver/directory_provisioner.go:
```diff
@@ -0,0 +1,135 @@
+		if err := d.mounter.MakeDir(target); err != nil {
+			return status.Errorf(codes.Internal, "Could not create dir %q: %v", target, err)
+		}
+		if err := d.mounter.Mount(fileSystemId, target, "efs", mountOptions); err != nil {
```
I did this is two separate defer statements so that each part had a chance to run at least even if one part failed, but I'm happy to change it if that's not what you meant

jonathanrainer commented on pkg/driver/directory_provisioner.go:
```diff
@@ -0,0 +1,135 @@
+	if err := d.mounter.MakeDir(target); err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not create dir %q: %v", target, err)
+	}
+	if err := d.mounter.Mount(fileSystemId, target, "efs", mountOptions); err == nil {
```
I tried to do this but `gofmt` kept putting it back and I couldn't get the form you stated here to compile. Could you give me a more concrete example for the `if` statement above?

jsafrane replied:
I meant, don't check for `err == nil`, check for `err != nil` and `return`. You won't need `else{}` branch at all and it becomes more straightforward.

jonathanrainer replied:
Oh of course, sorry, have done that now :)

andre-lx commented:
HI @jonathanrainer 

Great work. 

We are using the #640 for a while, but the 120 access points are a real blocker for us, so this looks like a better solution for our problem in the near future. 

We only have two problems that are solved by the #640 and that is the `subPathPattern`, since we want to write to the root folder, and we not want the creation of one folder per PVC. 

The other problem we see, is that, since you are not using access points, all the files in the EFS are written as root user. 

Why not use a single access point to do all the work? 

Thanks for the hard work and hope this two PR's are merged soon.

oe-hbk commented:
@jonathanrainer Thanks for this PR.  It solves a big problem for us where we can't use APs because we don't use Amazon DNS.  I took this and merged with https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/687  (fixing some merge conflicts along the way) and they both work great together.  One thing, which may or may not be related to the combined PR on my end, but for directory provisioning, when the SC reclaimPolicy: delete   the created directory does not get deleted.   If I set basePath: /foo/bar   when the PVC and PV are deleted, /foo/bar/pvc-unique-id   directory still remains.

FleetAdmiralButter commented:
Thanks for this @jonathanrainer! I've done a build of this and deployed to our cluster where I've force-updated the EFS storage class to use directory provisioning. Existing PVs all still mount correctly via access points and new PVs are being provisioned correctly using directories.

ChamMach commented:
Hello, any news on this? I think it's critical to merge it for this driver's future.

jsafrane commented:
@jonathanrainer can you please rebase the PR?
@wongma7  @torredil can you please review it? IMO it's quite solid.

jonathanrainer commented:
@jsafrane Sorry I've kind of let this one get away from me a bit, I've just changed job but should be able to dedicate some time to this later in the week. Think it needs a rebase and the E2Es need looking at but it's nearly there.

jonathanrainer commented:
Right @jsafrane this is now fixed, realised that I'd made a mistake in the implementation of `DeleteVolume` so the node was trying to delete the folder while it was still mounted and thus got stuck, which caused a knock on effect to other tests in the suite. Now that's fixed and we've got 🟢 E2Es, this is ready to be reviewed properly.

@wongma7 and @torredil if you could have a look I'd be really appreciative so we can push this towards getting merged whenever is convenient.

jsafrane commented on pkg/driver/directory_provisioner.go:
```diff
@@ -0,0 +1,144 @@
+		klog.V(5).Infof("Provisioning directory with permissions %s", perms)
+
+		provisionedDirectory := path.Join(target, provisionedPath)
+		err := d.osClient.MkDirAllWithPerms(provisionedDirectory, perms, uid, gid)
```
How does a pod that uses such a PV get access to the directory? In other words, how user or Kubernetes know that it should run it with allocated UID/GID? For AccessPoints, EFS itself makes sure that any UID/GID is translated to the allocated one.

IMO it would make more sense not to allocate any UID/GID at all and expect `DirectoryPerms: 777` in the storageClass or `fsGroup` in Pods.

jonathanrainer replied:
Yes I agree, I've made the change to make the permissions 777 by default but the options is still there in StorageClass to override them if people want to for whatever reason (I'm sure there are use-cases where it makes sense)

jsafrane replied:
The documentation is now kind of confusing now. I don't know how to make it better, maybe in a subsequent PR. It should say that for `efs-dir` it is recommended to have `DirectoryPerms: 777` and omit `gidRangeStart` / `gidRangeEnd`, so a pod can access its volume and on the other hand, `efs-ap` should have `DirectoryPerms: 770` and some `gidRangeStart`/`gidRangeEnd`.

All options are available in both modes if users really know what they are doing.

jonathanrainer replied:
Have made a change to the README to add our recommendations, let me know what you think

ChamMach commented:
Hello @jonathanrainer any news on this? 🤔

metasim commented:
This seems SO close be being done, and is SO desperately needed! 😄 Think this might make it over the finish line in the near term? 🙏

jonathanrainer commented on pkg/driver/directory_provisioner.go:
```diff
@@ -0,0 +1,144 @@
+		d.osClient.Remove(target)
+		return status.Errorf(codes.Internal, "Could not mount %q at %q: %v", fileSystemId, target, err)
+	}
+	if err := d.osClient.RemoveAll(target + subpath); err != nil {
```
So it is possible that root squash is enabled, but it's something user has to explicitly opt-into and you can assign roles in AWS such that the driver could circumvent that. It's probably worth adding some documentation for though as the IAM role would be different to standard, what do you think?

jsafrane replied:
Yes, please add some doc that `--delete-provisioned-dir` may not work when root_squash is used.

jonathanrainer replied:
Have added this to the README and the help for CLI args

jonathanrainer commented on pkg/driver/directory_provisioner.go:
```diff
@@ -0,0 +1,144 @@
+	}
+
+	defer func() {
+		if err := d.osClient.RemoveAll(target); err != nil {
```
Yes that makes sense, I've re-arranged the logic so it only tries to delete if the unmount is successful.

jsafrane replied:
It looks better now, thanks!

jonathanrainer commented on pkg/driver/controller.go:
```diff
@@ -77,221 +67,45 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
-			return nil, status.Errorf(codes.InvalidArgument, "File System does not exist: %v", err)
-		}
-		return nil, status.Errorf(codes.Internal, "Failed to fetch File System info: %v", err)
+		return nil, status.Errorf(codes.Internal, "Could not assign UID or GID to access point")
```
Done, you're right that's a lot clearer

jonathanrainer commented on pkg/driver/controller.go:
```diff
@@ -300,70 +114,17 @@ func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
 
 	//TODO: Add Delete File System when FS provisioning is implemented
```
Removed

jonathanrainer commented:
@jsafrane Hi, sorry it's taken so long but I've addressed your comments and rebased. Sadly I can't seem to work out a solution to the end-to-end tests, as every time I try and create the new folder the permissions set get ignored, hence the failure. Any ideas? I've been hitting my head against this for hours and can't find a solution :(

jsafrane commented:
@jonathanrainer I tried to run the tests, it seems to me that the error is caused by umask in the driver container. `cat /proc/1/status` shows `Umask:	0022` there and [MkDirAll](https://pkg.go.dev/os#MkdirAll) says:

> The permission bits perm (before umask) are used for all directories that MkdirAll creates.

So `w` bits are cleared from the group + other permissions. IMO it's not a bad idea to add `os.Chmod` after `os.MkDirAll`.

jonathanrainer commented:
@jsafrane Thanks so much, that got it sorted

@Ashley-wenyizha Realise the process of this merging might be a bit weird, so how can we progress this forward?

jsafrane commented:
I poked AWS people more than enough times. 
/lgtm
/approve

@jonathanrainer thanks a lot for the PR and your patience!

Ashley-wenyizha commented:
Hi @jonathanrainer 

Sorry for the delayed response, we will review this PR and see if we can release the feature in the coming releases.

Note : there are issues with our test cluster and we are doing a hot fix so `pull-aws-efs-csi-driver-external-test-eks` might fail on the PR.

Thanks for your contribution and sorry about the delayed response

jsafrane commented:
/retest

jonathanrainer commented:
> Hi @jonathanrainer
> 
> Sorry for the delayed response, we will review this PR and see if we can release the feature in the coming releases.
> 
> Note : there are issues with our test cluster and we are doing a hot fix so `pull-aws-efs-csi-driver-external-test-eks` might fail on the PR.
> 
> Meanwhile, I see we have 19 commits in the PR, could we rebase and merge these commits to few key commits?
> 
> Thanks for your contribution and sorry about the delayed response

Hi, no worries just glad we can get this merged eventually :) 

I've rebased and pulled it into 7 commits that cover the development process properly. Let me know if you want it more squashed or anything like that.

jonathanrainer commented:
/retest

jsafrane commented:
@Ashley-wenyizha so, are you going to review this? I am asking for review from AWS side for almost a year and always got a reply that "someone will look at it".

Ashley-wenyizha commented:
@jsafrane Yes my team will review this, sorry about the delay, there has been internal team changes upon efs-csi-driver so we can allocate better resources working on our open source projects and support our customers.

On top of the code review itself, AWS require security team's review on new features as well so that will also cause some delay.

Sorry again about the delay, we have now attention on this PR and https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/640

I will give update on this PR and PR 640 by the end of the week (5/12/2023) to see if we will need security review and timeline of this feature's release

These are very useful features and thanks for the effort and support @jonathanrainer!

jonathanrainer commented:
@Ashley-wenyizha is there any update on this in terms of timelines etc? Very happy to assist in this however I can

Ashley-wenyizha commented:
hi @jonathanrainer 

We are waiting for security team's reply, meanwhile we will start the code review and better understand of the new feature as we will need to write up the security survey as well for the new feature (and thanks for putting in the information in the feature overview as well, that is very helpful). 

Thanks for your patience and follow up!

Ashley-wenyizha commented:
@thesuperzapper, @andre-lx

Apologize for the super delayed response. We have started to support since Jan 17 2023

https://aws.amazon.com/about-aws/whats-new/2023/01/amazon-efs-increases-access-points-file-system/

Please feel free to ignore if you already aware of this feature enhancement.

jonathanrainer commented:
Hi @Ashley-wenyizha again, has there been any further progress or would you like any assistance in documenting the new feature? More than happy to help in any way I can :)

jonathanrainer commented:
@Ashley-wenyizha Hi, it's been over a month, do we have a timeline on a release plan for this and #640 ? Again, happy to help in any way I can :)

Ashley-wenyizha commented:
Hi @jonathanrainer, 

Sorry about the wait, we are still reviewing the PR internally and mostly in the queue of getting security team's bandwidth to review the changes. (together with https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/640)

I will give weekly update on this regarding the progress from now on and thanks again for your patience and contribution to the new feature.

jonathanrainer commented:
> Hi @jonathanrainer,
> 
> Sorry about the wait, we are still reviewing the PR internally and mostly in the queue of getting security team's bandwidth to review the changes & security risks. (together with #640)
> 
> I will give weekly update on this regarding the progress from now on and thanks again for your patience and contribution to the new feature.

No problem, thank you so much for continuing to push this forward. As I said happy to help in any way I can with this or #640

Ashley-wenyizha commented:
Hi Jonathan,

Giving you a quick weekly update on this PR, EFS security guardian was out of office this week, we are hoping to arrange meeting with our guardian next week.

Do you have any design doc/one pager you could share with us on this feature if you could forward us? that would help the guardian to easier understand the feature, design, security risks etc.

Meanwhile we just merged a PR from AWS doc team member, that might have introduced you some merge conflict, could you pull that change and fix the conflicts?

jonathanrainer commented:
> Hi Jonathan,
> 
> 1. Giving a quick weekly update on this PR, EFS security guardian was out of office this week, we are hoping to arrange meeting with our guardian next week.
> 
> * Do you have any design doc/one pager you could share with us on this feature if you could forward us? that would help the guardian to easier understand the feature, design, security risks etc.
> * Meanwhile we just merged a PR from AWS doc team member, that might have introduced you some merge conflict, could you pull that change and fix the conflicts?
> 
> 2. Update on PR640 - I left 1 comment on PR [Allow more control of the name of the directory created by the Access Point under Dynamic Provisioning #640](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/640), otherwise the implementation lgtm, If you could fix the conflict there (in PR#640) we can merge 640 first and then this PR can be updated afterwards.
> 
> thanks

Hi @Ashley-wenyizha have replied to your question on #640 and fixed up the changes after the docs change in both PRs. I don't have a one-pager right now but I'll write one and then post a link in this thread once that is complete :) Looking forward to getting these merged, there will be some non-trivial work to get the two PRs to work closely together as I've never tried that before but I'm sure it won't be too difficult, actually really looking forward to seeing these out in the wild!

Ashley-wenyizha commented:
Thanks Jonathan @jonathanrainer!

640 lgtm, could you only keep few commits (ideally 1-3) and we can merge 640 afterwards first. (almost there!)

Let me know when you have the one pager ready I will send it over along to efs security guardian

Ashley-wenyizha commented:
/retest

Ashley-wenyizha commented:
Ping to see any update here?

jonathanrainer commented:
@Ashley-wenyizha Hi, have rebased this but I know now the E2Es are failing (as is everything else), I'm not planning to do work to fix this yet because I want to wait for the other PR to merge anyway as I think that might be quite a complicated rebase as it is.

Once that's done I'll rebase and fix everything up at the same time and then it should be ready for review properly. Once that's done I'll post the one-pager as well and then it can all be reviewed together. Is that ok?

crackedupcorson commented:
Any update on when this feature might be available?

wolffberg commented:
Hi @jonathanrainer. Any chance you will have time for above in the near future? Your work is much appreciate⭐

wolffberg commented:
/remove-lifecycle stale

wolffberg commented:
/remove-lifecycle rotten

wolffberg commented:
/remove-lifecycle stale

z0rc commented:
/remove-lifecycle stale
